File: pounders/py/tests/regression_tests/test_failing_objective.py
# This tests the ability of the Python pounders implementation to
# gracefully terminate when receiving a NaN as a component in the vector of
# objective output.

import sys

import ibcdfo.pounders as pdrs
import numpy as np


def failing_objective(x):
    fvec = x

    if np.random.uniform() < 0.1:
        fvec[0] = np.nan

    return fvec


spsolver = 1
nfmax = 1000
gtol = 1e-13
n = 3
m = 3

X0 = np.array([10, 20, 30])
npmax = 2 * n + 1
L = -np.inf * np.ones(n)
U = np.inf * np.ones(n)
nfs = 0
F0 = []
xkin = 0
delta = 0.1
printf = 0

np.random.seed(1)

[X, F, flag, xk_best] = pdrs.pounders(failing_objective, X0, n, npmax, nfmax, gtol, delta, nfs, m, F0, xkin, L, U, printf, spsolver)
assert flag == -3, "No NaN was encountered in this test, but should have been."


F0 = np.array([1.0, 2.0])
nfs = 2
[X, F, flag, xk_best] = pdrs.pounders(failing_objective, X0, n, npmax, nfmax, gtol, delta, nfs, m, F0, xkin, L, U, printf, spsolver)
assert flag == -1, "We are testing proper failure of pounders"

File: pounders/m/tests/test_failing_objective.m
% This tests the ability of the MATLAB pounders implementation to
% gracefully terminate when receiving a NaN as a component in the vector of
% objective output.

function [] = test_failing_objective()

[here_path, ~, ~] = fileparts(mfilename('fullpath'));
oldpath = addpath(fullfile(here_path, '..'));

spsolver = 1;

nf_max = 1000;
g_tol = 1e-13;
n = 3;
m = 3;

xs = [10; 20; 30];
np_max = 2 * n + 1;
Low = -Inf(1, n);
Upp = Inf(1, n);
X_0 = xs';
delta_0 = 0.1;

rand('seed', 1);
Ffun = @(x)failing_objective(x);

[X, F, hF, flag, xk_best] = pounders(Ffun, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp);
assert(flag == -3, "No NaN was encountered in this test, but (with high probability) should have been.");

Model.spsolver = spsolver;

% Intentionally not passing a function for an objective
[X, F, hF, flag, xk_best] = pounders(X_0, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp, [], [], Model);
assert(flag == -1, "Should have failed");

% Intentionally putting a NaN in F to cover part of pounders.m
Ffun = @(x) nan(1, 3);
[X, F, hF, flag, xk_best] = pounders(Ffun, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp, [], [], Model);
assert(flag == -3, "Should have failed immediately after first eval");

% Intentionally given the wrong size of F to cover part of pounders.m
Ffun = @(x) ones(1, 6);
[X, F, hF, flag, xk_best] = pounders(Ffun, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp, [], [], Model);
assert(flag == -1, "Should have failed");

% Intentionally testing failures
didFail = false;
try
    [X, F, hF, flag, xk_best] = pounders(Ffun, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp, [], [], 'wrong');
catch ME
    didFail = true;    
    assert(strcmp(ME.message, 'Model must be a struct'));
end
assert(didFail, 'pounders call did not fail as expected');

didFail = false;
try
    [X, F, hF, flag, xk_best] = pounders(Ffun, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp, [], 'wrong');
catch ME
    didFail = true;    
    assert(strcmp(ME.message, 'Options must be a struct'));
end
assert(didFail, 'pounders call did not fail as expected');

didFail = false;
try
    [X, F, hF, flag, xk_best] = pounders(Ffun, X_0, n, nf_max, g_tol, delta_0, m, Low, Upp, 'wrong');
catch ME
    didFail = true;    
    assert(strcmp(ME.message, 'Prior must be a struct'));
end
assert(didFail, 'pounders call did not fail as expected');

path(oldpath);
